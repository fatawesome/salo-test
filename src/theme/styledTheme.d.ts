import 'styled-components';

declare module 'styled-components' {
  export interface DefaultTheme {
    boxShadow: string;
    colors: {
      background: '#FFFFFF';

      cta: string; // buttons, checkboxes, price (i.e. everything screaming on user)
      textPrimary: string;
      textSecondary: string;
    };
  }
}
