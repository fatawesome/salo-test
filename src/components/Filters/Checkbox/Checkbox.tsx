import React from 'react';
import styled from 'styled-components';

interface CheckboxProps {
  id: string;
  checked: boolean;
  onChange: React.ChangeEventHandler
  disabled?: boolean;
  className?: string
}

const Checkbox: React.FC<CheckboxProps> =
  ({ id,
     checked,
     disabled,
     children,
     onChange,
     className
  }) => {
  return (
    <div className={className}>
      <input
        type="checkbox" id={id}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
      />
      <label htmlFor={id}>
        <span data-testid="checkbox-box">
          {checked &&
            <img src={require('@/assets/icons/checkmark.svg')} alt="checked" />
          }
        </span>
        {children}
      </label>
    </div>
  )
};

const StyledCheckbox = styled(Checkbox)`
  &:hover {
    & * {
      cursor: pointer;
    }
    cursor: pointer;
    background-color: #F1FCFF;
  }
  
  input[type="checkbox"] {
    display: none;
    border: none !important;
    box-shadow: none !important;
  }

  input[type="checkbox"] + label span {
    width: 20px;
    height: 20px;
    vertical-align: middle;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border: 1px solid #9ABBCE;
    border-radius: 2px;
    margin-right: 10px;
  }

  input[type="checkbox"]:checked + label span {
    width: 20px;
    height: 20px;
    vertical-align: middle;
    border: 1px solid #2196F3;
  }
  
  label {
    display: block;
    font-weight: normal;
    font-size: 13px;
    line-height: 20px;
    padding: 10px 20px;
    width: 100%;
  }
`;

export default StyledCheckbox;
