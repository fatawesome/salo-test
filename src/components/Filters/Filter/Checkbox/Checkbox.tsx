import React from 'react';
import styled from 'styled-components';

interface CheckboxProps {
  id: string;
  checked: boolean;
  disabled?: boolean;
  onChange: React.ChangeEventHandler
}

const Checkbox: React.FC<CheckboxProps> =
  ({ id,
     checked,
     disabled,
     children,
     onChange
  }) => {
  return (
    <>
      <input
        type="checkbox" id={id}
        checked={checked}
        disabled={disabled}
        onChange={onChange}
      />
      <label htmlFor={id}>
        <span>
          {checked &&
            <img src={require('@/assets/icons/checkmark.svg')} alt="checked" />
          }
        </span>
        {children}
      </label>
    </>
  )
};

const StyledCheckbox = styled(Checkbox)`
  input[type="checkbox"] {
    display: none;
    border: none !important;
    box-shadow: none !important;
  }

  input[type="checkbox"] + label span {
    width: 20px;
    height: 20px;
    display: inline-block;
    vertical-align: middle;
  }

  input[type="checkbox"]:checked + label span {
    width: 20px;
    height: 20px;
    vertical-align: middle;
  }
  
  span {
    font-weight: normal;
    font-size: 13px;
    line-height: 20px;
  }
`;

export default StyledCheckbox;
