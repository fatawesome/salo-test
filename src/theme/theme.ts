import { DefaultTheme } from 'styled-components';

// TODO: font and stuff;
const aviasalesTheme: DefaultTheme = {
  boxShadow: '0px 2px 8px rgba(0, 0, 0, 0.1)',
  colors: {
    textPrimary: '#4A4A4A',
    textSecondary: '#A0B0B9',
    background: '#FFFFFF',
    cta: '#2196F3',
  }
}

export { aviasalesTheme };
