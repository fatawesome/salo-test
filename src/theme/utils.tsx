import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { aviasalesTheme } from './theme';
import { render } from '@testing-library/react';

const withTheme = (node: React.ReactNode) => (
  <ThemeProvider theme={aviasalesTheme}>
    {node}
  </ThemeProvider>
)

export const snapshotWithTheme = (node: React.ReactNode) => {
  return renderer.create(withTheme(node));
}

export const uiWithTheme = (node: React.ReactNode) => {
  return render(withTheme(node));
}

// TODO: можно обобщить в функцию вида (fn :: renderer -> node -> result)
