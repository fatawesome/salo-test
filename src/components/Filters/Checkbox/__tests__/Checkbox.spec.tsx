import React from 'react';

import Checkbox from '../Checkbox';
import { snapshotWithTheme, uiWithTheme } from '../../../../theme/utils';
import userEvent from '@testing-library/user-event';
import { queryHelpers, screen } from '@testing-library/react';

describe('Checkbox component', () => {
  const label = 'label'
  const changeHandler = jest.fn();
  let component: React.ReactNode;

  beforeEach(() => {
    changeHandler.mockClear();
    component = (
      <Checkbox
        id="test id"
        checked={true}
        onChange={changeHandler}
        disabled={false}
      >
        {label}
      </Checkbox>
    );
  })

  test('renders correctly', () => {
    const snapshot = snapshotWithTheme(component);
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('triggers change handler', () => {
    test('when clicked on text', () => {
      uiWithTheme(component);
      userEvent.click(screen.getByText(label));
      expect(changeHandler).toBeCalledTimes(1);
    });

    test('when clicked on box', () => {
      uiWithTheme(component);
      userEvent.click(screen.getByTestId('checkbox-box'));
      expect(changeHandler).toBeCalledTimes(1);
    })
  })
});
