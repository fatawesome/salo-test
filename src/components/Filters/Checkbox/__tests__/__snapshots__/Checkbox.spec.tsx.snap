// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component renders correctly 1`] = `
<div
  className="Checkbox__StyledCheckbox-sc-1nynq9x-0 kUavuH"
>
  <input
    checked={true}
    disabled={false}
    id="test id"
    onChange={[MockFunction]}
    type="checkbox"
  />
  <label
    htmlFor="test id"
  >
    <span
      data-testid="checkbox-box"
    >
      <img
        alt="checked"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </span>
    label
  </label>
</div>
`;
